### PR TITLE
Dataset and DataCollator for BERT Next Sentence Prediction (NSP) task.

### PR DESCRIPTION
**Pull request by mojave-pku**
I am working with some pre-training tasks recently, and found that transformers have not yet supported NSP tasks. 
So I write the code of **Dataset and DataCollator for BERT NSP task**, now I try to contribute.

This is my first contribution to an open source project. If there is something wrong, please point it out, thank you.

### CODE REVIEW

File: src/transformers/__init__.py [hunk 1]
```diff
@@ -442,12 +442,14 @@
         DataCollator,
         DataCollatorForLanguageModeling,
         DataCollatorForPermutationLanguageModeling,
+        DataCollatorForNSP,
```
Comment (JetRunner) [inline src/transformers/__init__.py:445]: `DataCollatorForNextSentencePrediction` is consistent with our naming patttern

**Review by JetRunner — COMMENTED** [review]
Hi Lianzhe, thanks for your contribution!

I wonder if there is a specific use case to train/fine-tune NSP alone? If I remember it right, BERT mixed the NSP objective with MLM objective.

File: src/transformers/__init__.py (hunk 1 above)
Comment (HuangLianzhe) [reply src/transformers/__init__.py:445]: Received, thanks for the reply.
The work I am doing recently involves  pre-training tasks using sentence-level only. Therefore, I write `DataCollatorForNSP` separately.
So do you suggest modifying directly on `DataCollatorForLanguageModeling` to add the codes for the NSP task? 
Is it necessary to keep the separate `DataCollatorForNSP`?

Comment (HuangLianzhe) [reply src/transformers/__init__.py:445]: I will commit the name modification later. `DataCollatorForNSP' -> `DataCollatorForNextSentencePrediction'

Comment (JetRunner) [reply src/transformers/__init__.py:445]: > So do you suggest modifying directly on `DataCollatorForLanguageModeling` to add the codes for the NSP task?
> Is it necessary to keep the separate `DataCollatorForNSP`?

No I'm not suggesting that! I only wonder if a separate `run_bert_nsp.py` (without MLM) script makes sense. What do you think? @sgugger @sshleifer

File: src/transformers/data/data_collator.py [hunk 2]
```diff
@@ -313,3 +314,136 @@ def mask_tokens(self, inputs: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor,
...
+class DataCollatorForNSP:
+    """
+    Data collator used for language modeling.
+    - collates batches of tensors, honoring their tokenizer's pad_token
+    - preprocesses batches for masked language modeling
+    """
+
+    tokenizer: PreTrainedTokenizer
+    mlm: bool = True
+    nsp_probability: float = 0.5
```
Comment (sgugger) [inline src/transformers/data/data_collator.py:329]: This is not used anywhere.

File: src/transformers/data/datasets/language_modeling.py [hunk 3]
```diff
@@ -99,3 +99,90 @@ def __len__(self):
 
     def __getitem__(self, i) -> torch.Tensor:
         return torch.tensor(self.examples[i], dtype=torch.long)
+
+
+class TextDatasetForNSP(Dataset):
```
Comment (sgugger) [inline src/transformers/data/datasets/language_modeling.py:104]: NSP should be NextSentencePrediction here.

File: src/transformers/data/data_collator.py [hunk 4]
```diff
@@ -313,3 +314,136 @@ def mask_tokens(self, inputs: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor,
...
+                )
+            return pad_sequence(examples, batch_first=True, padding_value=self.tokenizer.pad_token_id)
+
+    def create_examples_from_document(
+        self, document: List[List[int]], doc_index: int, examples: List[List[List[int]]]
+    ):
+        """Creates examples for a single document."""
+
+        # Account for [CLS], [SEP], [SEP]
+        max_num_tokens = self.block_size - 3
```
Comment (sgugger) [inline src/transformers/data/data_collator.py:369]: Shouldn't this be `block_size - tokenizer.num_special_tokens_to_add(pair=False)` to work with any tokenizer?

**Review by sgugger — CHANGES_REQUESTED** [review]
Thanks for the PR!

Even if it is your use case, I don't think it makes sense to support just the NSP task without the MLM objective. Especially since I don't think it's much work to add it to `NSPDataCollator` (just copy the part that masks the tokens from `DataCollatorForLanguageModeling`).

The script you added already has all the arguments for mlm so it then should just be renamed and it can both work for NSP-alone tasks such the ones you're studying (with mlm=0) and the actual BERT pretraining task.

What do you think?

File: src/transformers/data/datasets/language_modeling.py [hunk 5]
```diff
@@ -99,3 +99,90 @@ def __len__(self):
...
+
+                        # Empty lines are used as document delimiters
+                        if not line:
+                            self.examples.append([])
+                        tokens = tokenizer.tokenize(line)
+                        tokens = tokenizer.convert_tokens_to_ids(tokens)
+                        if tokens:
+                            self.examples[-1].append(tokens)
+
+                    # print(self.documents)
```
Comment (sgugger) [inline src/transformers/data/datasets/language_modeling.py:170]: Please clean up debugging code ;-)

**Comment by HuangLianzhe** [comment]
Thanks for reply and carefully review!
I understand what you and @JetRunner mean now. 
The purpose of submitting this script is to provide a tool that can test DataCollator and TextDataset (I don't know if I am doing this right). I agree with you and think it is not elegant to add a run_bert_nsp.py script alone. I will remove this script later and  support NSP together with MLM objective.
